core/domain/prod_validation_jobs_one_off.py

```diff
@@ -5038,6 +5039,98 @@ def _get_custom_validation_functions(cls):
         ]
 
 
+class PlaythroughModelValidator(BaseModelValidator):
+    """Class for validating PlaythroughModel."""
+
+    PLAYTHROUGH_PROJECT_RELEASE_DATETIME = datetime.datetime(2018, 9, 1)
+
+    @classmethod
+    def _get_external_id_relationships(cls, item):
+        return {
+            'exp_id': (
+                exp_models.ExplorationModel, [item.exp_id]),
+        }
+
+    @classmethod
+    def _get_model_id_regex(cls, unused_item):
+        return r'^[A-Za-z0-9-_]{1,%s}\.[A-Za-z0-9-_]{1,%s}$' % (
+            base_models.ID_LENGTH, base_models.ID_LENGTH)
+
+    @classmethod
+    def _get_model_domain_object_instance(cls, item):
+        return stats_services.get_playthrough_from_model(item)
+
+    @classmethod
+    def _validate_exploration_id_in_whitelist(cls, item):
+        """Validate the exploration id in playthrough model is in
+        the whitelist.
+
+        Args:
+            item: ndb.Model. PlaythroughModel to validate.
+        """
+        whitelisted_exp_ids_for_playthroughs = (
+            config_domain.WHITELISTED_EXPLORATION_IDS_FOR_PLAYTHROUGHS.value)
+
+        if item.exp_id not in whitelisted_exp_ids_for_playthroughs:
+            cls.errors['exploration id check'].append(
+                'Entity id %s: recorded in exploration_id:%s which '
+                'has not been curated for recording.' % (
+                    item.id, item.exp_id)
+            )
+
+    @classmethod
+    def _validate_reference(cls, item):
+        """Validate the playthrough is referenced by some issues of
+        the corresponding exploration.
+
+        Args:
+            item: ndb.Model. PlaythroughModel to validate.
+        """
+        exp_id = item.exp_id
+        exp_version = item.exp_version
+        exp_issues = (
+            stats_models.ExplorationIssuesModel.get_model(exp_id, exp_version))
+
+        has_reference_error = True
+        for exp_issue_dict in exp_issues.unresolved_issues:
+            if item.id in exp_issue_dict['playthrough_ids']:
```

Comment:
Let's make this check stronger:
-   There is exactly one `exp_issue` which holds the `playthrough_id`
-   That `exp_issue` must have the same `issue_type` as the playthrough.
-   That `exp_issue` must have the same value for `issue_customization_args[ISSUE_TYPE_KEYNAME_MAPPING[playthrough.issue_type]]`

Here's some psuedo-code to better describe these requirements:
```python
issues = []
for issue in exp_issues.unresolved_issues:
    if issue.issue_type == playthrough.issue_type:
        issue_customization_args = issue.issue_customization_args
        identifying_arg = (
            stats_models.ISSUE_TYPE_KEYNAME_MAPPING[
                issue.issue_type])
        if (issue_customization_args[identifying_arg] ==
                playthrough.issue_customization_args[identifying_arg]):
            issues.append(issue)
assert len(issues) == 1
assert playthrough_id in issues[0].playthrough_ids
```

Reply:
`assert len(issues) == 1` not only exert the firsts constraint

> There is exactly one `exp_issue` which holds the `playthrough_id`

, but also asserts there is only one issue that have the same `issue_type` and `identifying_arg`, is it intended? If not, I think the following snippet makes sense.

```python
issues = []
for issue in exp_issues.unresolved_issues:
    if issue.issue_type == playthrough.issue_type:
        issue_customization_args = issue.issue_customization_args
        identifying_arg = (
            stats_models.ISSUE_TYPE_KEYNAME_MAPPING[
                issue.issue_type])
        if (issue_customization_args[identifying_arg] ==
                playthrough.issue_customization_args[identifying_arg]):
            issues.append(issue)
count = 0
for issue in issues:
    if playthrough_id in issue.playthrough_ids:
        count += 1
assert count == 1
```

Reply:
Sure, that looks good to me. Let's also make sure exactly one `playthrough_id` is in the list.

Reply:
Done. I made some more change to the snippet I post that I think makes more sense than the previous two pseudocode, please take a look.